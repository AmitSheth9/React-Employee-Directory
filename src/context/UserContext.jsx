import { createContext, useContext, useState, useMemo } from "react";
import { getUser } from "../services/users";

const UserContext = createContext();

const UserProvider = ({ children }) => {
    const presentUser = getUser();
    const [user, setUser] = useState(presentUser ? { id: presentUser.id, email: presentUser.email } : {} );

    const [signedIn, setSignedIn] = useState(false);

    return <UserContext.Provider value={{ user, setUser, signedIn, setSignedIn }}>{children}</UserContext.Provider>
};

const useUser = () => {
    const context = useContext(UserContext);

    if (context === undefined) {
        throw new Error('useUser must be used inside UserProvider');
    }
    return context;
};

export { UserContext, UserProvider, useUser };