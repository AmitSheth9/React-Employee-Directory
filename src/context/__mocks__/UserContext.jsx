import { createContext, useContext, useState } from "react";


const UserContext = createContext();

const UserProvider = ({ mockUser, children }) => {
    
    const [user, setUser] = useState(mockUser ? { id: mockUser.id, email: mockUser.email } : {} );

    const [signedIn, setSignedIn] = useState(false);

    return <UserContext.Provider value={{ user, setUser, signedIn, setSignedIn }}>{children}</UserContext.Provider>
};

const useUser = () => {
    const context = useContext(UserContext);

    if (context === undefined) {
        throw new Error('useUser must be used inside UserProvider');
    }
    return context;
};

export { UserContext, UserProvider, useUser };