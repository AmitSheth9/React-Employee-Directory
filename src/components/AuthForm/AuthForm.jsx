import React from 'react'
import { useState, useEffect } from 'react';
import { useUser } from '../../context/UserContext';
import { signUpUser } from '../../services/users';
import { signInUser } from '../../services/users';
import { useHistory } from 'react-router-dom';
import { useLocation } from 'react-router-dom';
import { Link } from 'react-router-dom';

export default function AuthForm() {
    const auth = useUser();
    const history = useHistory();
    const [logIn, setLogIn] = useState(false);
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [errorx, setErrorX] = useState('');
    const location = useLocation();

    useEffect(() => {
        if (location.pathname === '/register' ) {
            setLogIn(false);
        }
        else if (location.pathname === '/login') {
            setLogIn(true);
        }
        
    }, [logIn])
    
    
    const handleSubmit = async (e) => {
        e.preventDefault();
      try {
        console.log(auth.user, 'submit');
        if(!email || password.length < 8 ) {
            throw new Error('Invalid attempt. Must be valid email and password of 8 characters')
        }
    } catch (error) {
      setErrorX(error.message);
      }
        
        if(location.pathname === '/register') {
            const returned = await signUpUser(email, password);
            console.log(returned);
            history.replace('/confirm-email');

        }
        if(location.pathname = '/login') {
            const userIn = await signInUser(email, password);
            console.log(userIn, 'userIn')
            auth.setUser({id: userIn.id, email: userIn.email});
            console.log(auth.user);
            history.replace('/profile')
        }
      
    }
    console.log(auth.user, 'outside');
    
    return (
        <div>
            <div>
            <form onSubmit={handleSubmit}>
                <fieldset>
                    <legend>{location.pathname==='/register' ? 'Create Account': 'Sign In'}</legend>
                    <label>Email:
                        <input 
                        value={email}
                        onChange={(e) => setEmail(e.target.value)}/> 
                    </label>
                    <label>Password:
                        <input 
                        value={password}
                        onChange={(e)=>setPassword(e.target.value)}/>
                    </label>
                    <button type='submit'>Submit</button>
                </fieldset>
            </form>
            <div>{location.pathname === '/register' ? <p>Already have an account?<Link to='/login'>Sign In</Link></p> : <p>Don't have an account? <Link to='/register'>Create an account here</Link></p>}
            </div>
        </div>
        </div>
    )
}
