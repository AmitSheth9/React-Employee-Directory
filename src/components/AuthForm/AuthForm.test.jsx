import {render, screen } from '@testing-library/react'
import { MemoryRouter, Switch } from 'react-router-dom';
import { UserProvider } from '../../context/UserContext';
import AuthForm from './AuthForm'
import reactRouterDom from 'react-router-dom';

jest.mock('../../services/users')
jest.mock('../../context/UserContext')

it('should render AuthForm', () => {
    const { container } = render(
        <UserProvider>
        
        <MemoryRouter>
            <AuthForm />
        </MemoryRouter>
     
        </UserProvider>
    );
    expect(container).toMatchSnapshot();
} );