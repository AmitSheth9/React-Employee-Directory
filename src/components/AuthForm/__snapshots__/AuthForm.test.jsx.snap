// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AuthForm 1`] = `
<div>
  <div>
    <div>
      <form>
        <fieldset>
          <legend>
            Sign In
          </legend>
          <label>
            Email:
            <input
              type="email"
              value=""
            />
          </label>
          <label>
            Password:
            <input
              type="password"
              value=""
            />
          </label>
          <button
            type="submit"
          >
            Submit
          </button>
        </fieldset>
      </form>
      <div />
      <div>
        <p>
          Don't have an account? 
          <a
            href="/register"
          >
            Create an account here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
